staging/src/k8s.io/apiserver/pkg/endpoints/request/requestcontext_test.go

```diff
@@ -0,0 +1,148 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package request
+
+import (
+	"net/http"
+	"sync"
+	"testing"
+)
+
+func TestRequestContextMapperGet(t *testing.T) {
+	mapper := NewRequestContextMapper()
+	context := NewContext()
+	req, _ := http.NewRequest("GET", "/api/version/resource", nil)
+
+	// empty mapper
+	if _, ok := mapper.Get(req); ok {
+		t.Fatalf("got unexpected context")
+	}
+
+	// init mapper
+	mapper.(*requestContextMap).init(req, context)
+	if _, ok := mapper.Get(req); !ok {
+		t.Fatalf("got no context")
+	}
+
+	// remove request context
+	mapper.(*requestContextMap).remove(req)
+	if _, ok := mapper.Get(req); ok {
+		t.Fatalf("got unexpected context")
+	}
+
+}
+func TestRequestContextMapperUpdate(t *testing.T) {
+	mapper := NewRequestContextMapper()
+	context := NewContext()
+	req, _ := http.NewRequest("GET", "/api/version/resource", nil)
+
+	// empty mapper
+	if err := mapper.Update(req, context); err == nil {
+		t.Fatalf("got no error")
+	}
+
+	// init mapper
+	if !mapper.(*requestContextMap).init(req, context) {
+		t.Fatalf("unexpected error, should init mapper")
+	}
+
+	context = WithNamespace(context, "default")
+	if err := mapper.Update(req, context); err != nil {
+		t.Fatalf("unexpected error")
+	}
+
+	if context, ok := mapper.Get(req); !ok {
+		t.Fatalf("go no context")
+	} else {
+		if ns, _ := NamespaceFrom(context); ns != "default" {
+			t.Fatalf("unexpected namespace %s", ns)
+		}
+	}
+}
+
+func TestRequestContextMapperConcurrent(t *testing.T) {
+	mapper := NewRequestContextMapper()
+
+	testCases := []struct{ url, namespace string }{
+		{"/api/version/resource1", "ns1"},
+		{"/api/version/resource2", "ns2"},
+		{"/api/version/resource3", "ns3"},
+		{"/api/version/resource4", "ns4"},
+		{"/api/version/resource5", "ns5"},
+	}
+
+	wg := sync.WaitGroup{}
+	for _, testcase := range testCases {
+		wg.Add(1)
+		go func(testcase struct{ url, namespace string }) {
+			defer wg.Done()
+			context := NewContext()
+			req, _ := http.NewRequest("GET", testcase.url, nil)
+
+			if !mapper.(*requestContextMap).init(req, context) {
+				t.Fatalf("unexpected init error")
+			}
+			if _, ok := mapper.Get(req); !ok {
+				t.Fatalf("got no context")
+			}
+			context2 := WithNamespace(context, testcase.namespace)
+			if err := mapper.Update(req, context2); err != nil {
+				t.Fatalf("unexpected update error")
+			}
+			if context, ok := mapper.Get(req); !ok {
+				t.Fatalf("got no context")
+			} else {
+				if ns, _ := NamespaceFrom(context); ns != testcase.namespace {
+					t.Fatalf("unexpected namespace %s", ns)
+				}
+			}
+		}(testcase)
+	}
+	wg.Wait()
+}
+
+func BenchmarkRequestContextMapper(b *testing.B) {
```

Comment:
run this both in master branch and here: the result below shows that we get  a little performance improvement
```
master

BenchmarkRequestContextMapper-8           500000              2951 ns/op
BenchmarkRequestContextMapper-8           500000              3009 ns/op
BenchmarkRequestContextMapper-8           500000              3022 ns/op
BenchmarkRequestContextMapper-8           500000              2933 ns/op
BenchmarkRequestContextMapper-8           500000              2946 ns/op


here 

BenchmarkRequestContextMapper-8           500000              2842 ns/op
BenchmarkRequestContextMapper-8           500000              2861 ns/op
BenchmarkRequestContextMapper-8           500000              2881 ns/op
BenchmarkRequestContextMapper-8           500000              2847 ns/op
BenchmarkRequestContextMapper-8           500000              2862 ns/op
```

Reply:
with higher parallelism (`b.SetParallelism(500)`), there is a much more noticeable impact:

```
master:
BenchmarkRequestContextMapper-8   	  500000	     10837 ns/op	     262 B/op	       8 allocs/op

PR:
BenchmarkRequestContextMapper-8   	  500000	      3898 ns/op	     295 B/op	      10 allocs/op
```